src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream abstraction over the TDS protocol which works to write the TDS packet
+    /// This stream will take care of wrapping the bytes with the TDS header. 
+    /// </summary>
+    internal class TdsWriteStream : Stream, ITdsWriteStream
+    {
+        private Stream _underlyingStream;
+
+        private byte[] _WriteBuffer;
+
+        // Start at the end of the Tds header
+        internal int WriteBufferOffset { get; private set; } = TdsEnums.HEADER_LEN;
+
+        // This should be set before flushing the buffer.
+        public TdsStreamPacketType PacketHeaderType { get; set; }
+
+        internal byte PacketNumber { get; private set; } = 1; // Packets always start with 1.
```

Comment:
Likewise, I don't think this needs to be internal and can be a private field. Can be easily tested by sending a packet (check for 1), send another packet (check for 2). Maybe even send a bunch of packets and make sure it rolls over.

Reply:
To add here, if `TdsWriteStream` is used in an unsynchronised context then `PacketNumber++` has a race condition and could send two packets with the same ID. It'd probably be cheapest to set the packet number to `Interlocked.Increment(ref myReference) & 0xFF`, but this requires the ability to pass a ref; a variable is usually the simplest way to handle this.

NB: the packet ID in the MARS packet header will need similar treatment. I've not reviewed on possible header size variations for MARS because I assume that a `TdsWriteStream` will be wrapped by a future `MarsStream` or similar.

Reply:
@benrr101  agree on the visibility.

@edwardneal what is the meaning of unsynchronised context? 
From the rest of your comment, I think you are hinting at concurrent execution. This stream will not be called concurrently. That's due to the way TDS protocol is. MARS is another story, and @benrr101  and I will be visiting MARS later. First we want to light up TDS.

Reply:
@saurabh500 - yes, that's correct and I agree that it eliminates the possible race condition.